### PR TITLE
Migrate Localization redirects to Cloudfront + Lambda

### PR DESCRIPTION
**Pull request by kylorhall**
[ticket](https://vimaly.com/#j8mqm/t/Migrate_localization_from_Javascript_into_AWS/zLS9aKu5deH9fnXmQ)

In AWS:
 - [Lambda](https://console.aws.amazon.com/lambda/home?region=us-east-1#/functions/localizeMarketingSiteCloudfrontRequests/versions/67)
 - [Cloudfront](https://console.aws.amazon.com/cloudfront/home?region=us-east-1#distribution-settings:E57A7O2HKWUL9) > Behaviors > Lambdas

Needs:
 - [x] Resolve or approve known issue (see ticket).
 - [x] Permissions resolved.
 - [x] Deploy resolved: or manual deploy..

### CODE REVIEW

File: scripts/localizationLambda.jest.js [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+  describe('isValidCountryCode', () => {
+    test('should return true when valid, regardless of case', () => {
+      expect.assertions(codesInAllCases.length);
+
+      codesInAllCases.forEach(code => {
+        expect(isValidCountryCode(code)).toEqual(true);
+      });
+    });
+
+    test('should return undefined when unknown', () => {
```
Comment (MichaelKuhinica) [inline scripts/localizationLambda.jest.js]: Should return false or undefined?

File: scripts/localizationLambda.jest.js [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+      expect(isValidCountryCode('EU')).toEqual(false);
+      expect(isValidCountryCode(1)).toEqual(false);
+      expect(isValidCountryCode({})).toEqual(false);
+      expect(isValidCountryCode(/whatever man/)).toEqual(false);
+    });
+  });
+
+  describe('getPathCountryCode', () => {
+    test('should return true when valid country code is found, regardless of case', () => {
+      expect(getPathCountryCode('/nz')).toEqual('nz');
```
Comment (MichaelKuhinica) [inline scripts/localizationLambda.jest.js]: Should return true or the normalised country code?

File: scripts/localizationLambda.jest.js [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+      badCodes.forEach(code => {
+        const request = { headers: { cookie: [{ value: `sharesight_country=${code}` }] } }
+
+        expect(getCookieCountryCode(request)).toEqual(undefined);
+      })
+    });
+  });
+
+  describe('pathWithoutCountryCode', () => {
+    test('should stip out the country code', () => {
```
Comment (MichaelKuhinica) [inline scripts/localizationLambda.jest.js]: stip => strip

File: scripts/localizationLambda.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  uri = uri.toLowerCase();
+
+  for (let i = 0; i < validCountryCodesLength; i++) {
+    const code = validCountryCodes[i];
+    if (
+      code &&
+      uri.indexOf(`/${code}`) === 0 && // eg. startsWith
+      (uri.length === 3 || // if only 3 characters, then it matches 100%
+        (uri.length > 3 && uri.indexOf('/', 1) === 3) // if 4+ length, it must have a trailing slash, eg. /nz/, not /nzd-something
+      )
```
Comment (MichaelKuhinica) [inline scripts/localizationLambda.js]: Nice! 👍

File: scripts/localizationLambda.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    ) {
+      return code; // return found country
+    }
+  }
+}
+
+function getHeaderCountryCode (request) {
+  let code = request.headers['cloudfront-viewer-country'][0].value;
```
Comment (MichaelKuhinica) [inline scripts/localizationLambda.js]: Is this structure guaranteed? I would add some safe guards just to be sure.

File: source/js/localization.js.es6 [hunk 6]
```diff
@@ -6,98 +6,103 @@
...
-        document.querySelectorAll(`a[href^="${config.help_url}"]`), // help site
-        document.querySelectorAll(`a[href^="${config.base_path}"]`), // relative urls
-      ].map(elements => Array.from(elements))
+  updateUrls () {
+    ;[].concat.apply([], // flatten arrays of arrays
+      [
+        config.base_url, // absolute urls (www.sharesight.com)
+        config.help_url, // help site (help.sharesight.com)
+        `${config.base_path}/`.replace(/\/+/g, '/'), // relative urls (/faq); replace duplicate slashes
+      ].map(path => Array.from(document.querySelectorAll(`a[href^="${path}"]`)))
```
Comment (MichaelKuhinica) [inline source/js/localization.js.es6]: That's a nice implementation! 👏

File: scripts/localizationLambda.js (hunk 5 above)
Comment (kylorhall) [reply scripts/localizationLambda.js]: I probably should guard everything at the top level as a failure would result in a horrid Cloudfront error page.  This is straight AWS code.

I'll just add a `isRequestValid`, `isResponseValid` type helper.  The problem is, if AWS changes this, we'll never know until we notice it's not redirecting.  I'd hope they never change this.

Comment (kylorhall) [reply scripts/localizationLambda.js]: Instead, I just wrapped this in a `try/catch` function... :|

File: lambda/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# Cloudfront + Lambda for localization
```
Comment (michael-sands) [inline lambda/README.md]: Seems pretty clear and easy to follow 👍

**Pull request merged**